### PR TITLE
Enable unused-awaitable error for mypy

### PR DESCRIPTION
**Pull request by MatMaul**
Since I got bitten already twice by this, let's see if this check makes sense.

This is to be reviewed on a per commit basis.
Each commit is handling a different type of ignorable reported error.

There are also a couple of commits that look more like small actual bugs than false positives.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [ ] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [ ] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by DMRobertson** [comment]
I think a few of these are functions which are decorated by `wrap_as_background_process`; I don't think it's a problem that their invocations are missing an await?

In general it seems like this has a fairly high false positive rate, which is sad to see.

**Comment by clokep** [comment]
> Do we want to do the same here or relying on the git history to find this info looks good enough ?
> 
> EDIT: I am planning to not squash this PR to keep the rationales around in the commit messages.

It looks like the git history just says "ignore these" -- I think squashing those all together would be fine.

I suspect it would be better to include the reasoning in comments in the code *or* add comments to `run_background_process` and friends to explain this.

**Comment by MatMaul** [comment]
> I think a few of these are functions which are decorated by wrap_as_background_process; I don't think it's a problem that their invocations are missing an await?

Indeed not sure why I didn't noticed. So no bugs, I'll include them in the relevant commit.

> It looks like the git history just says "ignore these" -- I think squashing those all together would be fine.

I should indeed add more details in the commit messages to explain the rationale of each type of changes.

> I suspect it would be better to include the reasoning in comments in the code or add comments to run_background_process and friends to explain this.

That makes complete sense for `run_background` type of ignores where I can indeed just comment those methods.

For https://github.com/matrix-org/synapse/pull/14519/commits/e39d7578b04883ac2fa2103f006b5171a79e92f1 it's however a bit more annoying since it is the same rationale but in quite a number of different places.
I could keep this commit separated, put a detailed explanation in the commit message and add a one liner `please see detailed explanation in the commit message` for each instance ?

**Comment by clokep** [comment]
@MatMaul Is this worth following-up with or should we close it?

**Comment by DMRobertson** [comment]
I've merged in develop and will fixup what I can.

File: tests/handlers/test_oidc.py [hunk 1]
```diff
@@ -287,7 +287,7 @@ def test_validate_config(self) -> None:
         """Provider metadatas are extensively validated."""
         h = self.provider
 
-        def force_load_metadata() -> Awaitable[None]:
+        def force_load_metadata() -> "OpenIDProviderMetadata":
```
Comment (DMRobertson) [inline tests/handlers/test_oidc.py:290]: This one is mine. I think this fix is correct(?)

File: tests/http/test_matrixfederationclient.py [hunk 2]
```diff
@@ -451,7 +451,7 @@ def test_client_does_not_retry_on_400_plus(self):
         self.failureResultOf(d)
 
     def test_client_sends_body(self):
-        defer.ensureDeferred(
+        defer.ensureDeferred(  # type: ignore[unused-awaitable]
```
Comment (DMRobertson) [inline tests/http/test_matrixfederationclient.py:454]: Should this have a `self.get_success` around it? Or is the `self.pump` below sufficient to ensure this gets called?

File: tests/storage/test_appservice.py [hunk 3]
```diff
@@ -297,7 +299,9 @@ def test_complete_appservice_txn_updates_last_txn_state(
         service = Mock(id=self.as_list[0]["id"])
         events = [Mock(event_id="e1"), Mock(event_id="e2")]
         txn_id = 5
-        self._set_state(self.as_list[0]["id"], ApplicationServiceState.UP)
+        self.get_success(
```
Comment (DMRobertson) [inline tests/storage/test_appservice.py:302]: Also not sure if this needs to be here? :shrug:

File: tests/util/caches/test_deferred_cache.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ class DeferredCacheTestCase(TestCase):
     def test_empty(self) -> None:
         cache: DeferredCache[str, int] = DeferredCache("test")
         with self.assertRaises(KeyError):
-            cache.get("foo")
+            cache.get("foo")  # type: ignore[unused-awaitable]
```
Comment (DMRobertson) [inline tests/util/caches/test_deferred_cache.py:29]: AFAICS  `cache` is roughly a clever `Mapping[str, Deferred]` throughout this file, and we're checking that we can get and set its contents. It's okay that we're not awaiting on them. But please double-check this.

**Review by DMRobertson — COMMENTED** [review]
I think this looks good, but I did do some of this myself---so would like another set of eyes.

AFAICS there are three categories of warnings we're suppressing here:

- `run_in_background` and `run_as_background_process` return deferreds so you can block until---or do something after---the background task completes.
- we call a method on a Deferred which returns `self`, and mypy complains that the return value is unused
- synchronous ops that return a Deferred (e.g. popping from a `Dict[str, Deferred]`)

The first class causes a lot of noise here. I wonder if we should make `run_in_background` and `run_as_background_process` return `None`---I don't _think_ we ever use their return values. (And if we did we could have two variants, e.g. `run_in_background_and_discard` versus `run_in_background_and_ which does return a deferred, say `run_in_background_returning_handler` or something.)

**Comment by squahtx** [comment]
> I think this looks good, but I did do some of this myself---so would like another set of eyes.
>
> AFAICS there are three categories of warnings we're suppressing here:
>
> - `run_in_background` and `run_as_background_process` return deferreds so you can block until---or do something after---the background task completes.
>
> The first class causes a lot of noise here. I wonder if we should make `run_in_background` and `run_as_background_process` return `None`---I don't _think_ we ever use their return values. (And if we did we could have two variants, e.g. `run_in_background_and_discard` versus `run_in_background_and_ which does return a deferred, say `run_in_background_returning_handler` or something.)

nb: we're supposed to await `run_in_background`, otherwise CPU usage accounting gets lost when the background task outlives the request that started it. We also get wonderful "Re-starting finished log context" log spam when that happens.

**Comment by DMRobertson** [comment]
> b: we're supposed to await `run_in_background`, otherwise CPU usage accounting gets lost when the background task outlives the request that started it. We also get wonderful "Re-starting finished log context" log spam when that happens.

Huh, thanks. That's news to me---the docstring at https://github.com/matrix-org/synapse/blob/d8cc86eff484b6f570f55a5badb337080c6e4dcd/synapse/logging/context.py#L797-L800 made it sound like it was kind-of-okay to not await `run_in_background`.

**Comment by DMRobertson** [comment]
> we're supposed to await run_in_background

Do you mean specifically using the `await` keyword here, or more generally awaiting by feeding it into Twisted's event loop?

Also, are the examples of using run_in_background in https://matrix-org.github.io/synapse/latest/development/synapse_architecture/cancellation.html#uncancelled-processing correct?

**Comment by squahtx** [comment]
> > we're supposed to await run_in_background
> 
> Do you mean specifically using the `await` keyword here, or more generally awaiting by feeding it into Twisted's event loop?

the `await` keyword

> Also, are the examples of using run_in_background in https://matrix-org.github.io/synapse/latest/development/synapse_architecture/cancellation.html#uncancelled-processing correct?

Good question. I think those are fine because there's an implicit wait for the task to finish (or at least hit the `to_resolve.callback(None)` line, after which there's no more async code).

... actually maybe those examples aren't okay. They'll certainly not re-start any finished logging contexts, but the active logging context will get marked as finished inside the `to_resolve.callback(None)` line, before it has been deactivated. Logging contexts are way too confusing.

File: synapse/_scripts/update_synapse_database.py [hunk 5]
```diff
@@ -59,7 +59,7 @@ async def run_background_updates() -> None:
 
     def run() -> None:
         # Apply all background updates on the database.
-        defer.ensureDeferred(
+        defer.ensureDeferred(  # type: ignore[unused-awaitable]
             run_as_background_process("background_updates", run_background_updates)
         )
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/_scripts/update_synapse_database.py:64]: I suspect that we can remove the `ensureDeferred` call.

File: synapse/util/async_helpers.py [hunk 6]
```diff
@@ -635,7 +635,7 @@ async def _ctx_manager() -> AsyncIterator[None]:
                 # writer waiting for us and it completed entirely within the
                 # `new_defer.callback()` call above.
                 if self.key_to_current_writer.get(key) == new_defer:
-                    self.key_to_current_writer.pop(key)
+                    self.key_to_current_writer.pop(key)  # type: ignore[unused-awaitable]
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/util/async_helpers.py:638]: > I wonder if `del self.key_to_current_writer.pop[key]` would work here?

Looks like it should?

> Is it possible for self.key_to_current_writer.get(key) to change between line 637 and 638? (I don't think so---there shouldn't be another thread accessing this?)

I don't think so either. The `ReadWriteLock` is designed to be async-safe, but not thread safe.

File: tests/http/test_matrixfederationclient.py (hunk 2 above)
Comment (squahtx) [reply tests/http/test_matrixfederationclient.py:454]: `ensureDeferred` will run the async function up until the first blocking await, ie. it should be fine.

File: tests/storage/test_appservice.py (hunk 3 above)
Comment (squahtx) [reply tests/storage/test_appservice.py:302]: Look reasonable!

**Review by squahtx — COMMENTED** [review]
In principle I like the idea of turning this check on. But there are an awful large number of false positives, which are easy to get fatigued by. I'm also fairly sure we shouldn't be discarding the result of `run_in_background`.

**Comment by MatMaul** [comment]
> I'm also fairly sure we shouldn't be discarding the result of `run_in_background`.

I am unsure what to do about that.
I think in most case where we don't currently use the result we want to switch to `run_as_background_process` ?
I am inferring that after reading the docstrings, since `run_as_background_process` explicitly mentions the case `(or for firing-and-forgetting in the middle of a normal synapse async function)`.

**Comment by squahtx** [comment]
> > I'm also fairly sure we shouldn't be discarding the result of `run_in_background`.
> 
> I am unsure what to do about that. I think in most case where we don't currently use the result we want to switch to `run_as_background_process` ?

That's probably the best way forward right now. Though I'd note that this has some implications regarding where resource usage is accounted under in metrics, and also makes it harder to determine which request kicked off which bit of processing.

**Comment by DMRobertson** [comment]
We've had this on our board for a while and I don't think we have the bandwidth to drive it forward. I'd _love_ to see this fixed and I think we've learned useful information from this, but I think we should park this for now.

If anyone gets time to look at this in the future, I'd suggest tackling one of the classes identified in https://github.com/matrix-org/synapse/pull/14519#pullrequestreview-1285868358  to hopefully make a smaller PR.

**Pull request closed without merging**